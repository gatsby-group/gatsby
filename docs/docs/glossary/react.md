---
title: React
---

Learn about React, including what it is, why you might use it, and how it fits into the Gatsby ecosystem.

## What is React?

React (sometimes written as React.js, or ReactJS) is a code library for building web sites and applications. It's written using [JavaScript](https://www.gatsbyjs.org/docs/glossary#javascript), one of the programming languages used to create web pages.

Facebook first released React in 2013. The company maintains the project, along with a community of contributors. React, like Gatsby, is free to use and open source under the terms of the [MIT License](https://github.com/facebook/react/blob/master/LICENSE).

React lets you create _components_ &mdash; reusable, shareable combinations of JavaScript, CSS, and HTML or SVG. You might, for example, build an `InputBox` component that includes an input field, label, and validation. You can then use the `InputBox` component on multiple pages, and across multiple projects. In other words, Components help you build web sites and single-page applications more quickly.

React is one of the foundational building blocks of Gatsby. Gatsby bundles React, [webpack](https://www.gatsbyjs.org/docs/glossary#webpack), [GraphQL](https://www.gatsbyjs.org/docs/glossary#graphql), as well as other libraries and tools into a single framework for building web sites. Rather than installing and configuring a development environment from scratch, you can install Gatsby and start building.

### Learn more about React

* [React](https://reactjs.org/) Official web site

