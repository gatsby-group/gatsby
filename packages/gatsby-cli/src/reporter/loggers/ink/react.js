const { createRequireFromPath } = require(`gatsby-core-utils`)

const inkRequire = createRequireFromPath(`ink`)

/* 
Make sure we use same "react" instance as "ink" would use to prevent error:

UNHANDLED REJECTION Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.

This should fix case 3. from that error.
/*/
module.exports = inkRequire(`react`)
