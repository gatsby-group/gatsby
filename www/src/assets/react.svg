<svg viewBox="0 0 47 42" xmlns="http://www.w3.org/2000/svg"><g fill="none" fill-rule="evenodd" transform="translate(1)"><g fill="#639"><path d="m30.5 35.454c-3.13 1.808-10.293-3.827-15.562-12.954-5.27-9.127-6.57-18.147-3.438-19.954 3.13-1.808 10.293 3.827 15.562 12.954 5.27 9.127 6.57 18.147 3.438 19.954zm1 1.733c4.522-2.61 3.07-12.682-2.706-22.687s-13.772-16.297-18.294-13.687-3.07 12.682 2.706 22.687 13.772 16.297 18.294 13.687z"/><path d="m0 19c0 5.22 9.448 9 21 9s21-3.78 21-9-9.448-9-21-9-21 3.78-21 9zm2 0c0-3.616 8.46-7 19-7s19 3.384 19 7-8.46 7-19 7-19-3.384-19-7z"/><path d="m28.794 23.5c5.776-10.005 7.228-20.076 2.706-22.687-4.522-2.61-12.518 3.682-18.294 13.687s-7.228 20.076-2.706 22.687c4.522 2.61 12.518-3.682 18.294-13.687zm-13.856-8c5.27-9.127 12.43-14.762 15.562-12.954 3.13 1.807 1.832 10.827-3.438 19.954s-12.43 14.762-15.562 12.954c-3.13-1.807-1.832-10.827 3.438-19.954z"/></g><rect fill="#ffb238" height="8" rx="4" width="8" x="17" y="15"/></g></svg>