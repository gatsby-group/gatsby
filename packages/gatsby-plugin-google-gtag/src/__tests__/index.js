import React from "react"
import { cleanup, fireEvent, render } from "@testing-library/react"
import { OutboundLink } from "../"

describe(`index.js`, () => {
  describe(`<OutboundLink />`, () => {
    afterEach(() => {
      cleanup()
      jest.resetAllMocks()
    })

    const setup = props => {
      const utils = render(<OutboundLink {...props}>link</OutboundLink>)

      return Object.assign({}, utils, {
        link: utils.getByText(`link`),
      })
    }

    it(`matches basic snapshot`, () => {
      const { container } = setup()

      expect(container).toMatchSnapshot()
    })

    describe(`tracking`, () => {
      beforeEach(() => {
        window.gtag = jest.fn()
      })
      it(`sends tracking event when clicked`, () => {
        const { link } = setup()

        fireEvent.click(link)

        expect(window.gtag).toHaveBeenCalled()
      })
      it(`sends custom tracking event when clicked`, () => {
        const eventCategory = `eventcategory`
        const eventAction = `eventaction`
        const eventLabel = `eventlabel`
        const eventValue = 55
        const { link } = setup({
          eventCategory,
          eventAction,
          eventLabel,
          eventValue,
        })

        fireEvent.click(link)

        expect(window.gtag).toHaveBeenCalledWith(
          `event`,
          eventAction,
          expect.objectContaining({
            event_category: eventCategory,
            event_label: eventLabel,
            event_value: eventValue,
          })
        )
      })
    })
  })
})
