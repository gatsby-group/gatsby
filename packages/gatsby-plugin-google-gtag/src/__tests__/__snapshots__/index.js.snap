// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.js <OutboundLink /> matches basic snapshot 1`] = `
<div>
  <a>
    link
  </a>
</div>
`;
